packages/core/core/src/AssetGraph.js

```diff
@@ -209,17 +211,53 @@ export default class AssetGraph extends Graph<AssetGraphNode> {
       return;
     }
 
-    let assetNodes = assets.map(asset => nodeFromAsset(asset));
-    this.replaceNodesConnectedTo(assetGroupNode, assetNodes);
+    let dependentAssetKeys = [];
+    let assetTuples: Array<[AssetNode, Array<Asset>, boolean]> = [];
+    for (let asset of assets) {
+      let isDirect = dependentAssetKeys.includes(asset.uniqueKey)
+        ? false
+        : true;
+
+      let dependentAssets = [];
+      for (let dep of asset.dependencies.values()) {
+        let dependentAsset = assets.find(
+          a => a.uniqueKey === dep.moduleSpecifier,
+        );
+        if (dependentAsset) {
+          dependentAssetKeys.push(dependentAsset.uniqueKey);
+          dependentAssets.push(dependentAsset);
+        }
+      }
+      assetTuples.push([nodeFromAsset(asset), dependentAssets, isDirect]);
+    }
 
-    for (let assetNode of assetNodes) {
-      let depNodes = [];
-      invariant(assetNode.type === 'asset');
-      for (let dep of assetNode.value.dependencies.values()) {
-        let depNode = nodeFromDep(dep);
-        depNodes.push(this.nodes.get(depNode.id) || depNode);
+    this.replaceNodesConnectedTo(
+      assetGroupNode,
+      assetTuples.filter(a => a[2]).map(a => a[0]),
```

Comment:
could these tuples be objects with named properties rather than relying on position?